### PR TITLE
elasticsearch: fix error responses

### PR DESCRIPTION
**Pull request by mhilton**

### CODE REVIEW

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/charmstore/2273/

File: elasticsearch/elasticsearch.go [hunk 1]
```diff
@@ -39,20 +39,75 @@ const (
...
+	return json.Unmarshal(b, &e.Reason)
 }
 
+// IsElasticsearchError reports whether the given error is an instance of
+// *ElasticsearchError.
+func IsElasticsearchError(err error) bool {
+	_, ok := err.(*ElasticsearchError)
+	return ok
+}
+
```
Comment (rogpeppe) [inline elasticsearch/elasticsearch.go:105]: Or perhaps:

     func ErrorStatusCode(err error) int {
         eserr, ok := err.(*ElasticSearchError)
         if ok {
              resurn eserr.Status
         }
         return 0
    }
?
then you can do:

    func IsConflictError(err error) bool {
       return ErrorStatusCode(err) == http.StatusConflict
    }

File: elasticsearch/elasticsearch.go [hunk 2]
```diff
@@ -123,48 +180,39 @@ func (db *Database) Alias(i, a string) error {
...
+// is not present, and returns a non-nil error if any other error occurs.
 func (db *Database) GetDocument(index, type_, id string, v interface{}) error {
 	d, err := db.GetESDocument(index, type_, id)
 	if err != nil {
-		return getError(err)
+		return errgo.Mask(err, IsElasticsearchError)
 	}
 	if !d.Found {
-		return ErrNotFound
+		return errgo.Mask(&ElasticsearchError{Status: http.StatusNotFound}, IsElasticsearchError)
```
Comment (rogpeppe) [inline elasticsearch/elasticsearch.go]: or errgo.WithCausef(nil, &ElasticSearchError{Status: http.StatusNotFound}, "")

File: elasticsearch/elasticsearch.go [hunk 3]
```diff
@@ -258,77 +306,70 @@ func (db *Database) PostDocument(index, type_ string, doc interface{}) (string,
...
 	index, type_, id string,
 	version int64,
 	versionType string,
 	doc interface{}) error {
 	url := fmt.Sprintf("%s?version=%d&version_type=%s", db.url(index, type_, id), version, versionType)
-	if err := db.put(url, doc, nil); err != nil {
-		return getError(err)
-	}
-	return nil
+	return errgo.Mask(db.put(url, doc, nil), IsElasticsearchError)
```
Comment (rogpeppe) [inline elasticsearch/elasticsearch.go:356]: You could potentially avoid all these explicit Mask calls by doing:


    var mask = errgo.MaskFunc(IsElasticSearchError)

then:

    return mask(db.put(url, doc, nil))

etc
*shrug*

File: elasticsearch/elasticsearch.go [hunk 4]
```diff
@@ -337,52 +378,49 @@ func (db *Database) Search(index, type_ string, q QueryDSL) (SearchResult, error
...
+	if err = json.Unmarshal(b, &errorBody); err != nil {
+		logger.Errorf("cannot unmarshal elasticsearch error: %s", err)
 	}
-	if v != nil {
-		if err = json.Unmarshal(b, v); err != nil {
-			log.Debugf("*** %s", err)
-			return errgo.Notef(err, "cannot unmarshal response")
-		}
+	// In some circumstances Not Found will be returned when it isn't an error, handle that.
+	if errorBody.Status == 0 && resp.StatusCode == http.StatusNotFound && v != nil {
```
Comment (rogpeppe) [inline elasticsearch/elasticsearch.go:424]: Do we really want to log the "cannot unmarshal" message in this case?

**Review by rogpeppe — APPROVED** [review]
LGTM with a few remarks.
Also, please state the problem that's being fixed in the commit message.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/charmstore/2274/

File: elasticsearch/elasticsearch.go [hunk 5]
```diff
@@ -39,20 +39,83 @@ const (
...
+
+type errorInfo struct {
+	Type   string `json:"type"`
+	Reason string `json:"reason"`
+}
+
+// UnmarshalJSON implements json.Unmarshaler
+func (e *errorInfo) UnmarshalJSON(b []byte) error {
+	type ei errorInfo
+	if b[0] == '{' {
```
Comment (fabricematrat) [inline elasticsearch/elasticsearch.go:83]: what if b is empty ?

**Review by fabricematrat — APPROVED** [review]

File: elasticsearch/elasticsearch.go (hunk 5 above)
Comment (mhilton) [reply elasticsearch/elasticsearch.go:83]: @rogpeppe tells me it won't be. The json package won't call it with an empty b.

**Comment by mhilton** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/charmstore-merge

**Pull request merged**